### DISCUSSION TITLE
[UI][QUESTION] Too much imports

### DISCUSSION THREAD

**Discussion by benjaminforras** [spartan-ng/spartan#66] [General]
Just a general idea and topic to talk about this.

While developing with the UI components I find myself importing a lot of brain directives and on top of them importing the respective helm directives, making the imports quite hard to maintain and keep track of.

Wouldn't it make sense to only import the helm directives? After all they are used with the brain directives.

Do someone have an idea how to make the imports not so cluttered? What do you guys think of this?

**Comment by goetzrobin** [comment]
Definitely a good point. I’m wondering if we should have the Helm components import their brain counterparts. I have adde module to improve the imports btw: BrnDialogModule & HlmDialogModule and so on will limit your imports to only two! Do you think that’s good enough? Do you think Hlm should be couple to brain via hostdirectives?

**Reply by benjaminforras** [reply]
> I have adde module to improve the imports btw

Modules are definetely an improvement, yes and I think it can be a workaround for now.

> Do you think Hlm should be couple to brain via hostdirectives?

I think so yes, however in this case we might need to reexport some of the inputs/outputs aswell so we can access them. Also I found out that some of the directives (for example `brnCtxMenuTriggerFor`'s `CdkContextMenuTrigger`) doesn't have their underlying outputs exported, so we cannot use them. (Going with the referenced `CdkContextMenuTrigger`, cannot bind the `(closed)`, `(opened)` events)

**Reply by elite-benni** [reply]
I also think that it is no problem to couple helm to brain.
I think I already did this in my PR (because i did not think of it as abproblem). 
It also makes implementation of hlm much easier in some cases.

**Reply by JoshuaNitschke** [reply]
I didn't see the modules mentioned in docs.  I was searching the discussion to see if anybody else was complaining about the number of imports - modules is a big improvement!

**Comment by tommyc38** [comment]
@goetzrobin I see that brain modules (e.g. calendar) exports an imports array (e.g. BrnCalendarImports) which is imported by the hlm-calendar component, for example.  This works but Webstorm is showing all kinds of errors in my hlm-calendar component template.  I can resolve this by importing each item individually.  How are you guys getting around this so you can actually leverage template linting for errors when using these type of imports?  I have tried using the spread operator but it still shows the brain directives in the template as out of scope (even though they work fine in storybook).

**Reply by ashley-hunter** [reply]
If I recall there were a few bugs in webstorm in this area (one similar one was https://youtrack.jetbrains.com/issue/WEB-61511/Support-standalone-component-pseudo-module-constants - which is supposed to have been fixed), but sometimes we get incorrect errors in webstorm as they don't use the Angular Language Service, they build their own implementation. 
So there's not a lot we can do, the code is perfectly valid, it's just Webstorm is reporting it as invalid.
